spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/NDVSketchGenerator.java

```diff
@@ -0,0 +1,176 @@
+
+    tuple2Iterator.forEachRemaining(tuple -> sketchMap.put(tuple._1, tuple._2));
+    return writeToPuffin(table, columnList, sketchMap);
+  }
+
+  private static StatisticsFile writeToPuffin(
+      Table table, List<String> columns, Map<String, ThetaSketchJavaSerializable> sketchMap)
+      throws IOException {
+    TableOperations operations = ((HasTableOperations) table).operations();
+    FileIO fileIO = operations.io();
+    String path = operations.metadataFileLocation(String.format("%s.stats", UUID.randomUUID()));
+    OutputFile outputFile = fileIO.newOutputFile(path);
+    try (PuffinWriter writer =
+        Puffin.write(outputFile).createdBy("Iceberg Analyze action").build()) {
+      for (String columnName : columns) {
+        writer.add(
+            new Blob(
+                StandardBlobTypes.APACHE_DATASKETCHES_THETA_V1,
+                ImmutableList.of(table.schema().findField(columnName).fieldId()),
+                table.currentSnapshot().snapshotId(),
+                table.currentSnapshot().sequenceNumber(),
+                ByteBuffer.wrap(sketchMap.get(columnName).getSketch().toByteArray()),
+                null,
+                ImmutableMap.of()));
+      }
+      writer.finish();
+
+      return new GenericStatisticsFile(
+          table.currentSnapshot().snapshotId(),
+          path,
+          writer.fileSize(),
+          writer.footerSize(),
+          writer.writtenBlobsMetadata().stream()
+              .map(GenericBlobMetadata::from)
+              .collect(ImmutableList.toImmutableList()));
+    }
+  }
+
+  static Iterator<Tuple2<String, ThetaSketchJavaSerializable>> computeNDVSketches(
+      SparkSession spark, String tableName, long snapshotId, List<String> columns) {
+    Dataset<Row> data =
+        spark
+            .read()
+            .option("snapshot-id", snapshotId)
+            .table(tableName)
+            .select(columns.stream().map(functions::col).toArray(Column[]::new));
+
+    final JavaPairRDD<String, String> pairs =
+        data.javaRDD()
+            .flatMap(
+                row -> {
+                  List<Tuple2<String, String>> columnsList =
+                      Lists.newArrayListWithExpectedSize(columns.size());
+                  for (int i = 0; i < row.size(); i++) {
+                    columnsList.add(new Tuple2<>(columns.get(i), row.get(i).toString()));
+                  }
+                  return columnsList.iterator();
+                })
+            .mapToPair(t -> t);
+
+    JavaPairRDD<String, ThetaSketchJavaSerializable> sketches =
+        pairs.aggregateByKey(
+            new ThetaSketchJavaSerializable(),
+            Integer.parseInt(
+                SQLConf.SHUFFLE_PARTITIONS().defaultValueString()), // number of partitions
+            (sketch, val) -> {
+              sketch.update(val);
+              return sketch;
+            },
+            new Combine());
+
+    return sketches.toLocalIterator();
+  }
+
+  static class Combine
```

Comment:
I think we can do this to get rid of ugly Function2 definition, and also make the main method a bit cleaner.

```
    JavaPairRDD<String, ThetaSketchJavaSerializable> sketches =
        pairs.aggregateByKey(
            new ThetaSketchJavaSerializable(),
            Integer.parseInt(
                SQLConf.SHUFFLE_PARTITIONS().defaultValueString()), // number of partitions
            NDVSketchGenerator::update,
            NDVSketchGenerator::combine);

    return sketches.toLocalIterator();
  }

  public static ThetaSketchJavaSerializable update(ThetaSketchJavaSerializable sketch, String val) {
    sketch.update(val);
    return sketch;
  }

    public static ThetaSketchJavaSerializable combine(
        final ThetaSketchJavaSerializable sketch1, final ThetaSketchJavaSerializable sketch2) {
      if (sketch1.getSketch() == null && sketch2.getSketch() == null) {
        return emptySketchWrapped;
      }
      if (sketch1.getSketch() == null) {
        return sketch2;
      }
      if (sketch2.getSketch() == null) {
        return sketch1;
      }

      final CompactSketch compactSketch1 = sketch1.getCompactSketch();
      final CompactSketch compactSketch2 = sketch2.getCompactSketch();
      return new ThetaSketchJavaSerializable(
          new SetOperationBuilder().buildUnion().union(compactSketch1, compactSketch2));
    }
```